File: 112-path-sum/112-path-sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    sm=0
    r=[]
    def hasPathSum(self, root, targetSum):
        """
        :type root: TreeNode
        :type targetSum: int
        :rtype: bool
        """
        if not root: return False
        
        if not root.left and not root.right:
            if root.val == targetSum:
                self.r.append(root.val)
                return True
            else:
                return False
        if self.hasPathSum(root.left,targetSum-root.val):
            self.r.append(root.val)
            return True
        if self.hasPathSum(root.right,targetSum-root.val):
            self.r.append(root.val)
            return True
        return False
                
        

File: 112-path-sum/112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
     List<List<Integer>>  res= new ArrayList<>();
    public boolean hasPathSum(TreeNode root, int targetSum) {
          if(root==null) return res.size()> 0;
        Stack<Integer> stk= new Stack<>();
        getPathSum(root,targetSum,stk);
        return res.size()> 0;
    }
    
 
    
   private void  getPathSum(TreeNode root, int targetSum,Stack<Integer> stk){
        stk.push(root.val);
       
       if(root.left==null && root.right==null && root.val==targetSum) res.add(new ArrayList<>(stk));
       if(root.left!=null)getPathSum(root.left,targetSum-root.val,stk);
       if(root.right!=null)getPathSum(root.right,targetSum-root.val,stk);
       
       stk.pop();
    }
}